server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -659,49 +659,94 @@ public IndexShardState markAsRecovering(String reason, RecoveryState recoverySta
 
     /**
      * Completes the relocation. Operations are blocked and current operations are drained before changing state to relocated. The provided
-     * {@link Runnable} is executed after all operations are successfully blocked.
+     * {@link BiConsumer} is executed after all operations are successfully blocked.
      *
-     * @param consumer a {@link Runnable} that is executed after operations are blocked
+     * @param consumer a {@link BiConsumer} that is executed after operations are blocked and that consumes the primary context as well as
+     *                 a listener to resolve once it finished
+     * @param listener listener to resolve once this method actions including executing {@code consumer} in the non-failure case complete
      * @throws IllegalIndexShardStateException if the shard is not relocating due to concurrent cancellation
      * @throws IllegalStateException           if the relocation target is no longer part of the replication group
-     * @throws InterruptedException            if blocking operations is interrupted
      */
-    public void relocated(final String targetAllocationId, final Consumer<ReplicationTracker.PrimaryContext> consumer)
-        throws IllegalIndexShardStateException, IllegalStateException, InterruptedException {
+    public void relocated(final String targetAllocationId,
+                          final BiConsumer<ReplicationTracker.PrimaryContext, ActionListener<Void>> consumer,
+                          final ActionListener<Void> listener)
+            throws IllegalIndexShardStateException, IllegalStateException {
         assert shardRouting.primary() : "only primaries can be marked as relocated: " + shardRouting;
         try (Releasable forceRefreshes = refreshListeners.forceRefreshes()) {
-            indexShardOperationPermits.blockOperations(30, TimeUnit.MINUTES, () -> {
-                forceRefreshes.close();
-                // no shard operation permits are being held here, move state from started to relocated
-                assert indexShardOperationPermits.getActiveOperationsCount() == OPERATIONS_BLOCKED :
-                        "in-flight operations in progress while moving shard state to relocated";
-                /*
-                 * We should not invoke the runnable under the mutex as the expected implementation is to handoff the primary context via a
-                 * network operation. Doing this under the mutex can implicitly block the cluster state update thread on network operations.
-                 */
-                verifyRelocatingState();
-                final ReplicationTracker.PrimaryContext primaryContext = replicationTracker.startRelocationHandoff(targetAllocationId);
-                try {
-                    consumer.accept(primaryContext);
-                    synchronized (mutex) {
+            indexShardOperationPermits.asyncBlockOperations(new ActionListener<>() {
```

Comment:
Also I think this was the last (non-test) usage of `IndexShardOperationPermits#blockOperations` so we can drop that method now.

Reply:
Right on both points, uff let me try and find a clean way of resolving the first point  :)

Reply:
How about this: https://github.com/elastic/elasticsearch/pull/64098/commits/9cdb46ba1c1d67af7ac555734d560cfc83578ff2 , this was the shortest change I could find to having the API we need for running the waiting under cancellable threads. Would also allow for cleanly removing `IndexShardOperationPermits#blockOperations` by using `SAME` for those in a follow-up.
Let me know what you think :)

Reply:
This seems rather to defeat the purpose of this change, since this is still a place that primary relocation still blocks.

Making `handoffPrimaryContext` async is a step in the right direction for sure, but IMO we should just continue to use `ISOP#blockOperations` here.

Reply:
> Making handoffPrimaryContext async is a step in the right direction for sure, but IMO we should just continue to use ISOP#blockOperations here.

We can't use the blocking API here if we want to async send the transport request because the blocking API does not give us access to the `Releasable` on the permits that we have to hold until we are done with `org.elasticsearch.indices.recovery.RecoveryTargetHandler#handoffPrimaryContext` => I figured this change is easier than refactoring the blocking variant. Maybe we can even find a way to not burn a thread for the acquiring of all the permits in some future follow-up, at least we already have a reasonable API for this with this approach?

Reply:
Aha I see, that makes sense then.